Sub-stakes description (#1558)
nucypher/nucypher: docs/source/architecture/slashing.rst

How slashing affects stake
--------------------------

.. _`Sub-stakes`: https://docs.nucypher.com/en/latest/architecture/sub_stakes.html

> The goal of slashing is to reduce the number of tokens that belongs to a staking offender. While the reduction of tokens is intended to be punitive, once the tokens have been slashed, the slashing algorithm attempts to preserve the most efficient use of the offenders' remaining tokens based on their existing `Sub-stakes`_. The main task is not to violate the logic of locking tokens.

Reviewer: "The main task is not to violate the logic of locking tokens."

Can you explain what you mean by this?

Author: we have such structure for staker in `StakingEscrow`
```
struct StakerInfo {
        uint256 value;
...
        SubStakeInfo[] subStakes;
    }
```
Technically slashing is just deacresing `value` field by slashing amount, because this `value` means how many tokens belong to staker. But what we call "slashing" here is to make all data still consistent.
If we just deacrese `value` and untouch `subStakes` array it could lead to such situation: sum if sub-stakes in some period in a future will be more than `value` and it means when staker will mint reward for that period he will get more tokens than he desirves. Also could be some other potential problems in the code but that is the main. So our task is to adjust values in  `subStakes` array

Reviewer 2: What about?

```suggestion
Slashing aims to reduce the overall number of tokens that belongs to a staking offender. While the reduction of tokens is intended to be punitive, once the tokens have been slashed, the slashing algorithm attempts to preserve the most efficient use of the offenders' remaining tokens based on their existing `Sub-stakes`_. The main goal is to decrease the tokens belonging to the staker while considering how tokens are allocated across different sub-stakes.
```

Author: I like it, what do you think @mswilkison ?

Reviewer: 👍

Author: ✔️

An entire stake consists of:

    * unlocked tokens which the staker can withdraw at any moment
    * tokens locked for a specific period

In terms of how the stake is slashed, unlocked tokens are the first portion of the stake to be slashed. After that, if necessary, locked sub-stakes are decreased in order based on their remaining lock time, beginning with the shortest. The shortest sub-stake is decreased, and if the adjustment of that sub-stake is insufficient to fulfil the required punishment sum, then the next shortest sub-stake is decreased, and so on. Sub-stakes that begin in the next period are checked separately.

> Sub-stakes for past periods cannot be slashed, so only the periods from now onward can be slashed. However, by design sub-stakes can't have a starting period that is after the next period, so all future periods after the next period will always have an amount of tokens less than or equal to the next period. The current period still needs to be checked since its stake may be different than the next period. Therefore, only the current period and the next period need to be checked for slashing.

Reviewer: ```suggestion
Sub-stakes for past periods cannot be slashed, so only the periods from the current period onward can be slashed. However, by design sub-stakes can't have a starting period that is after the next period, so all future periods after the next period will always have an amount of slashable tokens less than or equal to the next period. The current period still needs to be checked since its stake may be different than the next period. Therefore, only the current period and the next period need to be checked for slashing.
```

Author: Partially added, removed `slashable` form `amount of slashable tokens less than or equal to the next period` because it's important that all tokens in the stake less or equal to stake in the next period

    a. Calculate the maximum allowed total stake for any period for the staker ::

        max_allowed_stake = pre_slashed_total_stake - slashing_amount

       Therefore, for any period moving forward the sum of sub-stakes for that period cannot be more than ``max_allowed_stake``.
    b. For the current and next periods ensure that the amount of locked tokens is less than or equal to ``max_allowed_stake``. If not, then reduce the shortest sub-stake to ensure that this occurs; then the next shortest and so on, as necessary for the period.
    c. Since sub-stakes can extend over multiple periods and can only have a single fixed amount of tokens for all applicable periods (see `Sub-stakes`_), the resulting amount of tokens remaining in a sub-stake after slashing is the minimum amount of tokens it can have across all of its relevant periods. To clarify, suppose that a sub-stake is locked for periods ``n`` and ``n+1``, and the slashing algorithm first determines that the sub-stake can have 10 tokens in period ``n``, but then it can only have 5 tokens in period ``n+1``. In this case, the sub-stake will be slashed to have 5 tokens in both periods ``n`` and ``n+1``.

>     d. The above property of sub-stakes means that there is the possibility that the total amount of locked tokens for a particular period could be reduced to even lower than the ``max_allowed_stake`` when unnecessary. Therefore, the slashing algorithm may create new sub-stakes on the staker's behalf to utilize tokens in the earlier period, when a sub-stake is needed to be reduced to an even lower value because of the next period. In the example above in c), the sub-stake was reduced to 5 tokens because of period ``n+1``, so there are 5 "extra" tokens `(10 - 5)` available in period ``n`` that can still be staked; hence, a new sub-stake with 5 tokens would be created to utilize these tokens in period ``n``. This benefits both the staker, by ensuring that their remaining tokens are efficiently utilized, and the network by maximizing its health.

Reviewer: ```suggestion
    d. The above property of sub-stakes means that there is the possibility that the total amount of locked tokens for a particular period could be reduced to even lower than the ``max_allowed_stake``. Therefore, the slashing algorithm may create new sub-stakes on the staker's behalf to utilize tokens in the earlier period, when a sub-stake is needed to be reduced to an even lower value because of the next period. In the example above in c), the sub-stake was reduced to 5 tokens because of period ``n+1``, so there are 5 "extra" tokens `(10 - 5)` available in period ``n`` that can still be staked; hence, a new sub-stake with 5 tokens would be created to utilize these tokens in period ``n``. This benefits both the staker, by ensuring that their remaining tokens are efficiently utilized, and the network by maximizing its health.
```

Author: ✔️

To reinforce the algorithm, consider the following example stake and different slashing scenarios:

**Example:**

    A staker has 1000 tokens:
        * 1st sub-stake = 500 tokens locked for 10 periods
        * 2nd sub-stake = 200 tokens for 2 periods

>         * 3rd sub-stake = 100 tokens locked starting from the next period and locked for 5 periods. The 3rd sub-stake is locked for the next period but has not yet been used as a deposit for "work" - not until the next period begins.

Reviewer: ```suggestion
        * 3rd sub-stake = 100 tokens locked starting from the next period and locked for 5 periods. The 3rd sub-stake is locked for the next period but won't be used as a deposit for "work" until the next period begins.
```

Author: ✔️

>     * tokens which the staker can withdraw at any moment

Reviewer 2: ```suggestion
    * unlocked tokens which the staker can withdraw at any moment
```

Author: ✔️

    * tokens locked for a specific period

.. _`Sub-stakes`: https://docs.nucypher.com/en/latest/architecture/sub_stakes.html

> `Sub-stakes`_ get slashed in the order of their remaining lock time, beginning with the shortest – so the first portion of the stake to be slashed is the unlocked portion. After that, if necessary, locked sub-stakes are decreased – the shortest sub-stake is decreased by the required amount; if the adjustment of that sub-stake is insufficient to fulfil the required punishment sum, then the next shortest sub-stake is decreased, and so on. Sub-stakes that begin in the next period are checked separately.

Author: Will be slashed shortest sub-stake - which starts from tomorrow to minimize influence of slashing

> `Sub-stakes`_ get slashed in the order of their remaining lock time, beginning with the shortest – so the first portion of the stake to be slashed is the unlocked portion. After that, if necessary, locked sub-stakes are decreased – the shortest sub-stake is decreased by the required amount; if the adjustment of that sub-stake is insufficient to fulfil the required punishment sum, then the next shortest sub-stake is decreased, and so on. Sub-stakes that begin in the next period are checked separately.

Reviewer 2: Should we use our "locking duration" term here instead of "lock time"?

Author: For explanation of slashing we don't use strict periods, just comparison between sub-stakes, so I see that it's ok to write something simple like "lock for" or "lock time". But again if we want to have strict termenology everywhere - I'll change this

Reviewer 2: That's fair - changing it might introduce more confusion than necessary.

**Example:**

    A staker has 1000 tokens:

>         * 1st sub-stake = 500 tokens locked for 10 periods

Reviewer 2: Should we use "locking duration"...? here and in other spots in the document?
```suggestion
        * 1st sub-stake = 500 tokens with a locking duration of 10 periods
```

Reviewer 3: What about this comment from @derekpierre ?

Author: Actually that's resolved, see one resolved comment above

    Result:

        * 1st sub-stake = 500 tokens locked for 10 periods
        * 2nd sub-stake = 200 tokens for 2 periods
        * 3rd sub-stake = 100 tokens locked starting from the next period

>         * 100 tokens in an unlocked state

Author: I'm not sure I agree with you.
I see this as help to reader with what to expect from the next scenarios

* *Scenario 2*: Staker incurs penalty calculated to be worth **300 tokens**:

>     The unlocked tokens can only cover 200 tokens. In the current period, 700 tokens are locked and 800 tokens are locked for the next period. Therefore, we should reduce amount of locked tokens for the next period and leave unchanged locked amount in the current period. The 3rd sub-stake suits for this purpose but it's not the shortest one. So we take the 2nd sub-stake (the shortest), reduce it to 100 tokens and add new sub-stake with 100 tokens which active only in the current period.

Author: I like your suggestion except one sentence `In this scenario, the next period's sum is reduced and the current period's sum remains unchanged.` Which related to your questions.
Before slashing staker has 1000 tokens, fine is 300 tokens, so after slashing maximum locked tokens must be equal or less 700. It's why current period is untouched

>     The unlocked tokens can only cover 200 tokens worth of the penalty. Beyond that, the staker has 700 tokens currently locked and 100 tokens that will lock in the next period, meaning 800 tokens will be locked in total. In this scenario, we should reduce amount of locked tokens for the next period and leave unchanged locked amount in the current period. The 3rd sub-stake would be suitable to be reduced except that it's not the shortest, in terms of its unlock date. Instead, the 2nd sub-stake – the shortest (2 periods until unlock) – is reduced to 100 tokens and a new sub-stake with 100 tokens is added which is only active in the current period.

Reviewer: "Instead, the 2nd sub-stake – the shortest (2 periods until unlock) – is reduced to 100 tokens and a new sub-stake with 100 tokens is added which is only active in the current period."

Wait... why is a new sub-stake added? This leaves us with 800 tokens when there should only be 700 remaining.

Author: when we reduce 2nd sub-stake to 100 then we have 600 locked in current period and 700 in the next one, so to reestablish current period we create new sub-stake

Reviewer 2: @vzotova thanks for adding the intermediate steps. Let's add a little explanation to the last diagram for Scenario 2 explaining our creation of the 4th sub-stake. You can reuse some words from my attempt (😊 ) of an explanation in https://github.com/vzotova/nucypher/pull/5.

Similar kind of thing for Scenario 3 i.e. add an explanation that the 2nd sub-stake is emptied, but gets replenished for one period. So we didn't "create" a new sub-stake, but we did reorganize tokens so that the 2nd sub-stake had tokens 1 period.

   
* *Scenario 3*: Staker incurs penalty calculated to be worth **400 tokens**:

>     The difference from the previous scenario is that should also decrease locked tokens in the current period. At the first step the 2nd sub-stake is reduced to 100 tokens. Next step - adjustment for the next period. The shortest sub-stake still the same - the 2nd. And we need to deacrese it from 100 to 0 only for the next period. Will be the same if we change duration of the 2nd sub-stake from 2 periods to 1 and the other sub-stakes remain unchanged.

Author: ✔️

>     The difference between this and the previous scenario is that the current period's sum of locked tokens is also reduced. The first step is to reduce the 2nd sub-stake to 100 tokens. Then, the next period is adjusted – the shortest sub-stake is still the 2nd – and it is reduced from 100 to zero for the next period. Notably, this would have the same result if we changed the duration of the 2nd sub-stake from 2 periods to 1 and the other sub-stakes remained unchanged.

Reviewer: I think the way we're explaining this is very confusing. We have 1,000 initial tokens and in this and the last example, we slash X amount but the `Result` displays an amount (1,000 - X + 100) which I think will confuse everyone

Author: Could you point where is 1,000 - X + 100?

Reviewer 2: @vzotova , it's the fact that we start with 1000 tokens, get slashed 400 tokens, but there are 700 (500 + 100 + 100) tokens accounted for in our `Result` statement.

It's unclear why the 2nd sub-stake still ends up with 100 tokens...? even though the full 200 tokens are slashed form it.

Reviewer 2: Is a timing thing, where the 3rd sub-stake has 100 tokens that won't be locked until the next period - so the same 100 tokens are instead staked for 1 period as part of the 2nd sub-stake?

Author: Yep, both current and next period will have 600 locked tokens, 
I was counting on the picture to help explain with this moment

Reviewer 2: We'll need to explain this timing thing a more clearly. The diagram helps, but still needs a solid explanation as a lead in.

Author: For me it's difficult to rephrase this, too many time I work with slashing :)
Also keep in mind that ideally user will read "slashing" after "sub-stakes"

Reviewer 2: Can we add something general to the beginning explanation of our slashing methodology?

Basically what I've gleaned from our discussion is that we make every attempt to limit idle "not yet locked" tokens after slashing; these idle tokens will be used in earlier periods before their original lock time, to supplement a slashed locked stake to attempt to keep the pre-slashed total locked stake the same for earlier periods.

Assuming the above is correct, perhaps we can add something like:

```
Once a stake is slashed, tokens that are not currently locked but are set to be locked during future periods, i.e. currently "idle", can be used to replenish/supplement a slashed stake in an attempt to maintain the original pre-slashed total locked tokens for a period.  These "idle" tokens can only be used in periods before their starting lock period, and if necessary, a separate sub-stake is created to facilitate this action.
```

In our scenarios that involve the timing situation above (namely Scenario 2 & 3), we should indicate the available "idle" tokens that are then used to supplement slashed stakes, and the new sub-stake if created.

Author: I'm so confused with this explanation, I found somehow wrong some ideas from it. 
Next try :)
First I whant to point that stake is time function

I'll try to explain abstraction and how I was thinking when wrote code:
we have initial figure that consists of several sub-stakes + unlocked tokens, when we slash we move imaginary line from the top to bottom and this line "destroys" or "presses" sub-stakes starting from the shortest, but we can have sub-stakes only as rectangles so when imaginary line pushes - whole sub-stake becomes smaller

next step is restoring some sub-stakes where possible to minimize damage (time-issues in explanation come from this part): we still have this imagimary line that destroyed some sub-stakes and stoped but we will add some small sub-stakes which do not cross this line and restores some locks - this source for 4th sub-stake in 2nd scenario. 
In 3rd scenario imaginary line works as just press, 3rd sub-stake "stronger" then "second" (because longer) and 3rd "cuts" part of 2nd under the press. 

Not sure that'll help :)

Reviewer 2: I think we are in line with the understanding, seems like just figuring out a wording that works 😅 .

In Scenario 2: 
the 2nd sub-stake is reduced from 200 to 100 tokens, and the 100 "idle" tokens from the 3rd sub-stake and used to create the 4th sub-stake for staking for 1 period.

In Scenario 3:
The 2nd sub-stake is emptied from 200 to 0 tokens, and the 100 "idle" tokens from the 3rd sub-stake are used in the 2nd sub-stake directly (since it has no tokens) for 1 period.

Correct?

Author: can't say that is correct 
There is no such thing like "idle" tokens
Maybe this picture will show why I'm so confused:
```
            stake
            ^
            |
         700+-----+----+
            |     | 3rd|
         600|     +----+-------------+
            |    2nd   |     3rd     |
         500+----------+-------------+----------+
            |                                   |
            |               1st                 |
            |                                   |   period
            +-----------------------------------+--->
```

Author: I've added more diagrams in slashing doc

 
* *Scenario 4*: Staker incurs penalty calculated to be worth **600 tokens**:

>     Reducing the unlocked remaining tokens, 3rd sub-stakes, and the shortest sub-stake (2nd) is not enough, so they are all removed. The next shortest sub-stake is the 1st which is reduced from 500 to 400.

Author: ✔️
